src/Common/MemoryTracker.cpp

```diff
@@ -172,65 +172,81 @@ void MemoryTracker::alloc(Int64 size)
+        }
 #endif
 
-    std::bernoulli_distribution fault(fault_probability);
-    if (unlikely(fault_probability && fault(thread_local_rng)) && memoryTrackerCanThrow(level, true))
-    {
-        /// Prevent recursion. Exception::ctor -> std::string -> new[] -> MemoryTracker::alloc
-        BlockerInThread untrack_lock(VariableContext::Global);
-
-        ProfileEvents::increment(ProfileEvents::QueryMemoryLimitExceeded);
-        const auto * description = description_ptr.load(std::memory_order_relaxed);
-        amount.fetch_sub(size, std::memory_order_relaxed);
-        throw DB::Exception(DB::ErrorCodes::MEMORY_LIMIT_EXCEEDED,
-                            "Memory tracker{}{}: fault injected. Would use {} (attempt to allocate chunk of {} bytes), maximum: {}",
-                            description ? " " : "", description ? description : "",
-                            formatReadableSizeWithBinarySuffix(will_be),
-                            size, formatReadableSizeWithBinarySuffix(current_hard_limit));
-    }
-
-    if (unlikely(current_profiler_limit && will_be > current_profiler_limit))
-    {
-        BlockerInThread untrack_lock(VariableContext::Global);
-        DB::TraceCollector::collect(DB::TraceType::Memory, StackTrace(), size);
-        setOrRaiseProfilerLimit((will_be + profiler_step - 1) / profiler_step * profiler_step);
-    }
+        std::bernoulli_distribution fault(fault_probability);
+        if (unlikely(fault_probability && fault(thread_local_rng)) && memoryTrackerCanThrow(level, true))
+        {
+            /// Prevent recursion. Exception::ctor -> std::string -> new[] -> MemoryTracker::alloc
+            BlockerInThread untrack_lock(VariableContext::Global);
+
+            ProfileEvents::increment(ProfileEvents::QueryMemoryLimitExceeded);
+            const auto * description = description_ptr.load(std::memory_order_relaxed);
+            amount.fetch_sub(size, std::memory_order_relaxed);
+            throw DB::Exception(DB::ErrorCodes::MEMORY_LIMIT_EXCEEDED,
+                                "Memory tracker{}{}: fault injected. Would use {} (attempt to allocate chunk of {} bytes), maximum: {}",
+                                description ? " " : "", description ? description : "",
+                                formatReadableSizeWithBinarySuffix(will_be),
+                                size, formatReadableSizeWithBinarySuffix(current_hard_limit));
+        }
 
-    std::bernoulli_distribution sample(sample_probability);
-    if (unlikely(sample_probability && sample(thread_local_rng)))
-    {
-        BlockerInThread untrack_lock(VariableContext::Global);
-        DB::TraceCollector::collect(DB::TraceType::MemorySample, StackTrace(), size);
-    }
+        if (unlikely(current_profiler_limit && will_be > current_profiler_limit))
+        {
+            BlockerInThread untrack_lock(VariableContext::Global);
+            DB::TraceCollector::collect(DB::TraceType::Memory, StackTrace(), size);
```

Comment:
But profiling should remain enabled in "light" version of MemoryTracker.

Reply:
@alexey-milovidov  DB::TraceCollector::collect can throw exception.

Reply:
Let's make it non throwing.

Reply:
Checked DB::TraceCollector::collect does not throw exception, it use special buffer subclass that does not throw.